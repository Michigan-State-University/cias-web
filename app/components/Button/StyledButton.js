import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';
import { colors, themeColors } from 'theme';
import {
  margin,
  border,
  layout,
  padding,
  flex,
} from 'components/BaseComponentStyles';

const invertedStyles = color => css`
  background-color: ${colors.white};
  color: ${themeColors[color]};
  border: 1px solid ${themeColors[color]};
`;

const basicStyles = (outlined, color) => css`
  background-color: ${outlined ? colors.white : themeColors[color]};
  color: ${colors.white};
`;

const getHover = props => {
  if (props.hoverable && !props.disabled) {
    if (props.inverted) return basicStyles(props.outlined, props.color);
    return invertedStyles(props.color);
  }
  return '';
};

export const StyledButton = styled.button`
  width: 100%;
  height: 40px;
  border-radius: 100px;
  cursor: pointer;
  border: none;
  outline: none;
  ${props =>
    props.inverted
      ? invertedStyles(props.color)
      : basicStyles(props.outlined, props.color)};
  ${props => props.disabled && `cursor: default; background-color: grey;`};
  transition: background-color 300ms ease, color 300ms ease, border 300ms ease;
  &:hover {
    ${props => getHover(props)};
  }
  ${margin};
  ${border};
  ${layout};
  ${padding};
  ${flex};
`;

StyledButton.propTypes = {
  color: PropTypes.oneOf(['primary', 'secondary']),
  outlined: PropTypes.bool,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  disabled: PropTypes.bool,
  hoverable: PropTypes.bool,
};

StyledButton.defaultProps = {
  color: 'primary',
  disabled: false,
  hoverable: false,
};
