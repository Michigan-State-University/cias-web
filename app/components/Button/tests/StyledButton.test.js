import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import { StyledButton } from '../StyledButton';

describe('<StyledButton />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(<StyledButton />);
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render styled button and match the snapshot', () => {
    const { container } = render(<StyledButton />);
    expect(container).toMatchSnapshot();
  });

  it('Should render inverted button and match the snapshot', () => {
    const { container } = render(<StyledButton disabled inverted />);
    expect(container).toMatchSnapshot();
  });
});
