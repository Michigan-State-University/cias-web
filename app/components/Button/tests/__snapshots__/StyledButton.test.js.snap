// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StyledButton /> Should render inverted button and match the snapshot 1`] = `
.c0 {
  width: 100%;
  height: 40px;
  border-radius: 100px;
  cursor: pointer;
  border: none;
  outline: none;
  background-color: #FFFFFF;
  color: #107969;
  border: 1px solid #107969;
  cursor: default;
  color: #9AA1B6;
  border-color: #9AA1B6;
  -webkit-transition: background-color 300ms ease,color 300ms ease,border 300ms ease;
  transition: background-color 300ms ease,color 300ms ease,border 300ms ease;
}

<div>
  <button
    class="c0"
    color="primary"
    disabled=""
    radius="100px"
  />
</div>
`;

exports[`<StyledButton /> Should render styled button and match the snapshot 1`] = `
.c0 {
  width: 100%;
  height: 40px;
  border-radius: 100px;
  cursor: pointer;
  border: none;
  outline: none;
  background-color: #107969;
  color: #FFFFFF;
  -webkit-transition: background-color 300ms ease,color 300ms ease,border 300ms ease;
  transition: background-color 300ms ease,color 300ms ease,border 300ms ease;
}

<div>
  <button
    class="c0"
    color="primary"
    radius="100px"
  />
</div>
`;
